pandas/core/arrays/masked.py

```diff
@@ -217,6 +218,18 @@ def copy(self: BaseMaskedArrayT) -> BaseMaskedArrayT:
         mask = mask.copy()
         return type(self)(data, mask, copy=False)
 
+    @doc(ExtensionArray.factorize)
+    def factorize(self, na_sentinel: int = -1) -> Tuple[np.ndarray, ExtensionArray]:
+        arr = self._data
+        mask = self._mask
+
+        codes, uniques = _factorize_array(arr, na_sentinel=na_sentinel, mask=mask)
```

Comment:
is there a reason you want to call a private routine like this directly? shouldn't factorize just handle this directly? (isn't that the point of _values_for_factorize).

Reply:
This is the way we also do it in the base EA `factorize` method. 
The reason we are using this, and not `pd.factorize` directly, is because the public factorize does not support the additional `na_value` and `mask` keywords. 

The point of `_values_for_factorize` is indeed to avoid that EA authors have to call this private `_factorize_array` method themselves (and to make it easier to implement `EA.factorize`), but here, I explicitly do not use the general `_values_for_factorize` path to be able to customize/optimize the `IntegerArray/BooleaArray.factorize()` method specifically for those dtypes.

Reply:
this is really polluting the interface, I would much rather just add they keywords. It seems we are special casing EA to no end. This needs to stop.

> The reason we are using this, and not pd.factorize directly, is because the public factorize does not support the additional na_value and mask keywords.

Reply:
Expanding the public interface of `factorize` is out of scope for this PR, IMO. The implementation I put here above is *exactly* how we already do it for 2 years (we *are already* using `_factorize_array` in our other EAs) . If you want to do a proposal to change this, please open an issue to discuss.